Homework #1

iSergius opened the pull request:
Задания выполнены все. Но не судите строго это мой первый проект.
Хотелось бы получить feedback о моем коде. Какие есть недочеты и дурные привычки.

Commit:
First version

iSergius committed:
Correct method mutate

iSergius committed:
Change names virables

iSergius committed:
Change names variables

iSergius committed:
Correct mutate method

iSergius committed:
Correct counters

iSergius committed:
Add check value percent in method mutate

p-matvienko commented on osx-project-1/osx-project-1/Mutator.m:
```diff
@@ -0,0 +1,54 @@
+    int count3 = 0;
+    int count4 = 0;*/
+    
+    for (int i = 0;i<percent; i++) {
```
лучше было бы проценты перевести в количество заменяемых ячеек :)

iSergius replied:
В метод mutate мы передаем значение типа integer (-2,147,483,648 – +2,147,483,647) Мы по условию задачи не можем передать дробное число. Соответственно нам нужно только проверить значение percent на соответствию диапазону от 0...100, что я и исправил в последнем комите.

paillechat commented on osx-project-1/osx-project-1/Cell.m:
```diff
@@ -0,0 +1,62 @@
+            
+            int randomNumber = (int)(arc4random() % 4);
+            
+            NSString *memberDnaArray = NULL;
```
Можно было все гены из которых состоит ДНК сложить в один массив, а потом из него случайным образом получать элемент.
Это будет выглядеть более рационально и понятно.

Пример:

_types = [NSArray arrayWithObjects: @"A", @"T", @"G", @"C", nil];
for (i=0; i < 100; i++) {
  // получаем случайный ген
   [_DNA addObject: _types[arc4random() % 3]];
}

paillechat commented on osx-project-1/osx-project-1/Cell.m:
```diff
@@ -0,0 +1,62 @@
+    return self;
+}
+
+- (int) hammingDistance:(Cell *)insertCell {
```
В заголовочном файле Cell.h мето hammingDistance объявлен как - (int) hammingDistance: (Cell*) dna; а имплементирован как - (int) hammingDistance:(Cell *)insertCell, как же в итоге входящая переменная зовется?)

iSergius replied:
Это я пропустил когда переименовывал переменные. Хотя XCode не заругался так как важно что бы совпадали типы передаваемого значения.  Поправлю в ближайшем комите.

paillechat commented on osx-project-1/osx-project-1/Mutator.m:
```diff
@@ -0,0 +1,59 @@
+
+#import "Mutator.h"
+
+@implementation Cell (mutator)
```
Выглядит странным получение нового гена из массива DNA. 

Думаю было бы более логично завести массив со всеми возможными генами и реализовать метод "getAnotherGet: currentGen" который возвращает ген отличающийся от гена на выбранной позиции.

В остальном все имеет право на жизнь. Только устарелый код (который закомментирован) лучше убирать, если хочешь сохранить старые версии программы используй контроль версий, делай комиты по чаще и всегда будет возможность вернуться к тому что было.

igolovin01 commented on osx-project-1/osx-project-1/Cell.m:
```diff
@@ -0,0 +1,62 @@
+        
+        DNA = [NSMutableArray arrayWithCapacity:100];
+        
+        for (int i=0; i<100; i++) {
```
Для значния длины массива вместо цифры 100 лучше создать отдельную переменную с понятным именем и комментарием. В случае если с вашим кодом будет работать кто-то другой это значительно упростит ему жизнь. Придерживайтесь этого правила по всему коду, найдите все свои цифры 100 и замените их на одну переменную.

igolovin01 commented on osx-project-1/osx-project-1/Cell.m:
```diff
@@ -0,0 +1,62 @@
+        NSString *memberSelfArray = [selfArrayDna objectAtIndex:i];
+        NSString *memberGivenArray = [givenArrayDna objectAtIndex:i];
+        
+        if (memberSelfArray != memberGivenArray) {
```
Для сравнения значений строк в NSString существет специальный метод isEqualToString, используйте его вместо !=

igolovin01 commented on osx-project-1/osx-project-1/Cell.h:
```diff
@@ -0,0 +1,17 @@
+#import <Foundation/Foundation.h>
+
+@interface Cell : NSObject {
+    NSMutableArray *DNA;
```
У вас одновременно обьявлены и переменная NSMutableArray *DNA и @property NSMutableArray *DNA, нужно выбрать что-то одно и работать только с ним.

slebedev replied:
Это почему еще?

igolovin01 replied:
На мой взгляд это делает код менее читабильным, запутывает других разработчиков и усложняет отладку приложения.

slebedev replied:
Ну, на вкус и цвет, как говорится... Ничего плохого в этом нет, а иногда - даже нужно, чтобы было написано именно так. Свойство - автоматически создаваемый интерфейс для какой-либо переменной, так что как раз сразу видно, что есть переменная, и у нее есть геттеры и сеттеры, с ней можно работать, не проверяя, есть ли соответствующая реализация. 

Вы должны понимать, что если в интерфейсе видите 

@property NSMutableArray *DNA;

это автоматически означает, что на самом деле написано что-то вроде этого: 

{
    NSMutableArray *_DNA; // название зависит от synthesize
}
@property NSMutableArray *DNA;
-(void)setDNA:(NSMutableArray *)DNA;
-(NSMutableArray *)DNA;

Да и в учебных материалах свойства учат именно так, используя совместно с переменными класса.

iSergius committed:
- Add Read only property;
- Correct name insert Cell object in method hammingDistance;
- OOP Implementation method init and hammingDistance

iSergius committed:
v2 FINAL
Redesign implementation method mutate
Optimization init
